### PR TITLE
Add age column for all resources when using 'kubect get'

### LINKED ISSUE #8929: kubectl get should show creation time for other resources
e.g. being able to know when a namespace was created (at a  glance) would be nice.

### PR DESCRIPTION
**Pull request by hurf**
Fixes #8929

Dispaly human readable creation time like 'age' column do in 'kubectl
get po' for all resources.

### CODE REVIEW

**Comment by thockin** [comment]
ok to test

**Comment by satnam6502** [comment]
As far as I can tell kubectl get pods already reports an age?

**Comment by satnam6502** [comment]
Ah, RCs don't.

**Comment by hurf** [comment]
@satnam6502 It seems I don't have the permission to access the e2e result. It says `Access not possible`. Can you paste the log where the error is located here?

**Comment by jlowdermilk** [comment]
@hurf, the e2e run timed out on cluster initialization. I believe the reason is that the addition of an AGE column on `kubectl get nodes` output broke the `validate-cluster.sh` script [here](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/validate-cluster.sh#L43).

**Comment by jlowdermilk** [comment]
Oh and regarding e2e result, the [build log](https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/81a4e145ae12806e35107d83e0ed4cdb18327363/kubernetes-pull-build-test-e2e-gce/567/build-log.txt) that k8s-bot linked to above has the relevant output.

**Comment by derekwaynecarr** [comment]
So how many columns until get pods is unreadable again? :-)

**Comment by hurf** [comment]
@derekwaynecarr - Fortunately this patch hasn't add any column to `get po`, maybe we will worry about  `get rc` later:-)

**Comment by hurf** [comment]
@jlowdermilk - Thanks I'll look into it.

**Comment by hurf** [comment]
@jlowdermilk - To avoid the error may be caused by other column changes, I modified the script to find the STATUS column number instead of changing resource printing order or assigning column number directly in the script. But I'm not good at shell script, can you review the code again? I'll be happy to have your advice.

**Comment by jlowdermilk** [comment]
LGTM.

Bash is not my strong suit either, but your changes lgtm. The shippable failure looks like a break at head that you unluckily pulled in. Try rebasing to fix.

**Comment by wojtek-t** [comment]
@hurf - can you please rebase your PR?

**Comment by bgrant0607** [comment]
@derekwaynecarr One solution to lots of columns is `-o wide`: #10262. I don't think this PR cross the threshold, though.

**Comment by bgrant0607** [comment]
FYI @andronat

**Comment by bgrant0607** [comment]
@hurf Please rebase. Your tree looks like it's in an inconsistent state:
```
# github.com/GoogleCloudPlatform/kubernetes/test/integration
_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/integration/service_account_test.go:257: createdPod.Spec.ServiceAccount undefined (type "github.com/GoogleCloudPlatform/kubernetes/pkg/api".PodSpec has no field or method ServiceAccount)
_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/integration/service_account_test.go:258: createdPod.Spec.ServiceAccount undefined (type "github.com/GoogleCloudPlatform/kubernetes/pkg/api".PodSpec has no field or method ServiceAccount)
FAIL	github.com/GoogleCloudPlatform/kubernetes/test/integration [build failed]
```
I'd like to get this into 1.0.

**Comment by bgrant0607** [comment]
LGTM otherwise

**Comment by andronat** [comment]
@hurf please rebase to latest version as this #9876 PR that added extra optional label columns and changed they way printing is made. You might have to heavily refactor.

**Comment by hurf** [comment]
Sorry everybody, I was on a trip in previous days, I'll rebase it ASAP.

**Comment by hurf** [comment]
I didn't do anything to commit 1916018 except rebase, but e2e produced 2 different results. Also shippable reported failure. Anyone can paste the test result here for me to trace the error?

File: pkg/kubectl/resource_printer.go [hunk 1]
```diff
@@ -600,7 +606,7 @@ func printEndpointsList(list *api.EndpointsList, w io.Writer, withNamespace bool
 }
 
 func printNamespace(item *api.Namespace, w io.Writer, withNamespace bool, columnLabels []string) error {
-	if _, err := fmt.Fprintf(w, "%s\t%s\t%s\n", item.Name, formatLabels(item.Labels), item.Status.Phase); err != nil {
+	if _, err := fmt.Fprintf(w, "%s\t%s\t%s\t%s\n", item.Name, formatLabels(item.Labels), item.Status.Phase, translateTimestamp(item.CreationTimestamp)); err != nil {
```
Comment (andronat) [inline pkg/kubectl/resource_printer.go]: Can you please delete the last `\n`? It shouldn't be there from the beginning. Thanks

**Comment by jlowdermilk** [comment]
@hurf, shippable failure looks like flake (test-integration.sh timeout). Kicked.

**Comment by hurf** [comment]
@andronat - done. @jlowdermilk - Thanks! can you review it again?

File: cluster/validate-cluster.sh [hunk 2]
```diff
@@ -32,15 +32,24 @@ attempt=0
...
+  #     NAME                     LABELS    STATUS   AGE
+  #     kubernetes-minion-03nb   <none>    Ready    10m
   #
   # Echo the output, strip the first line, then gather 2 counts:
   #  - Total number of nodes.
   #  - Number of "ready" nodes.
   "${KUBE_ROOT}/cluster/kubectl.sh" get nodes > "${MINIONS_FILE}" || true
   found=$(cat "${MINIONS_FILE}" | sed '1d' | grep -c .) || true
-  ready=$(cat "${MINIONS_FILE}" | sed '1d' | awk '{print $NF}' | grep -c '^Ready') || true
+
```
Comment (bgrant0607) [inline cluster/validate-cluster.sh]: Not your fault, but this is not a good idea. Scripts should:
1. specify --api-version, and
2. use -o template

The human-oriented output format may change at any time.

cc @fabioy

Comment (hurf) [inline cluster/validate-cluster.sh]: @bgrant0607 - I was tring to reduce the influence caused by format change. And I think your opinion is more reasonable. I'll try to fix it in your way.

Comment (jlowdermilk) [inline cluster/validate-cluster.sh]: Good point, this should really have been using `-o template` in the first place (and would then have been unaffected by your change). The following command should be useful as a starting point:
```
kubectl get nodes -o template --template='{{range .items}}{{$cond := index .status.conditions 0}}{{$cond.type}},{{$cond.status}}:{{end}}' --api-version=v1
```

Comment (hurf) [inline cluster/validate-cluster.sh]: @jlowdermilk - Thanks for the hint. But I'm not going to use '--template' flag in the command for it is bound with api-version. I'm planning to use only '--api-version' and dump all output, then for different version, we can define different logic. I think it is good for further extension, maybe we'll have v2 or v3 later.

**Comment by bgrant0607** [comment]
LGTM

Restarted Shippable

cc @deads2k @ghodss

**Comment by bgrant0607** [comment]
Assigned to @thockin for another LGTM, since he requested this feature

**Comment by bgrant0607** [comment]
Please run `gofmt -s`. hack/verify-gofmt.sh will tell you which files.

Install presubmit hooks:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/development.md#hooks

**Comment by hurf** [comment]
@bgrant0607 - Sorry I don't quite catch you. Does the hook related to the shippable failure?

**Comment by jlowdermilk** [comment]
@hurf, yes the hook would catch the current shippable failure at commit time.

**Comment by hurf** [comment]
@jlowdermilk - Still confusing. The hook looks like a gofmt check, I didn't know it is related to shippable. Anyway I'll have a try to see what would happen.

**Comment by bgrant0607** [comment]
@hurf Shippable runs the checks. Sorry for the confusion.

**Comment by hurf** [comment]
@bgrant0607 - So you mean the shippable failure for this patch is cause by gofmt check? I'll look into it.

**Comment by hurf** [comment]
@bgrant0607 - BTW, it's inconvenient that developers don't have permission to view shippable results. Hope the problem can be solved later.

**Comment by bgrant0607** [comment]
Re. Shippable results: I agree. #9502

**Comment by wojtek-t** [comment]
It seems that this PR is ready for final review:
@bgrant0607 - IIUC, you already LGTMed it in https://github.com/GoogleCloudPlatform/kubernetes/pull/10008#issuecomment-115848688
@thockin - can you please take a look?

**Comment by bgrant0607** [comment]
Or maybe @roberthbailey, since there are changes to validate-cluster.sh.

**Comment by bgrant0607** [comment]
@hurf Please squash after you get another LGTM

**Comment by hurf** [comment]
@bgrant0607 - I think the two commits are doing different things. They're not strongly bound to each other. So better to keep them seperated. What do you think?

**Comment by hurf** [comment]
@thockin - If you have time, can you review this pr?

File: cluster/validate-cluster.sh [hunk 3]
```diff
@@ -24,23 +24,27 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
...
-  #     kubernetes-minion-03nb   <none>    Ready
-  #
-  # Echo the output, strip the first line, then gather 2 counts:
+  # Echo the output and gather 2 counts:
   #  - Total number of nodes.
   #  - Number of "ready" nodes.
-  "${KUBE_ROOT}/cluster/kubectl.sh" get nodes > "${MINIONS_FILE}" || true
-  found=$(cat "${MINIONS_FILE}" | sed '1d' | grep -c .) || true
-  ready=$(cat "${MINIONS_FILE}" | sed '1d' | awk '{print $NF}' | grep -c '^Ready') || true
+  "${KUBE_ROOT}/cluster/kubectl.sh" get nodes --api-version="${API_VERSION}" -o yaml> "${MINIONS_FILE}" || true
```
Comment (roberthbailey) [inline cluster/validate-cluster.sh]: Why does this still use the shell script? I thought we had been transitioning to just using kubectl directly?

File: cluster/validate-cluster.sh [hunk 4]
```diff
@@ -24,23 +24,27 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
...
+  
+  found=0;ready=0
+  if [[ $API_VERSION == v1 ]]; then
+    # when there is a 'kind: Node' line, there is a node found.
+    found=$(cat "${MINIONS_FILE}" | grep -c 'kind: Node') || true
+    # only when status: "True" and type: Ready both exist in conditions section, the node is "ready".
+    ready=$(cat "${MINIONS_FILE}" | sed -n 'N;/status: "True".*type: Ready/p; D;' | grep -c 'True') || true
```
Comment (roberthbailey) [inline cluster/validate-cluster.sh]: Have you tested this sed command on both linux and mac (we've seen a lot of cases where the mac version of sed doesn't work as well as the linux version)?

File: cluster/validate-cluster.sh [hunk 5]
```diff
@@ -24,23 +24,27 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
...
+  # Echo the output and gather 2 counts:
   #  - Total number of nodes.
   #  - Number of "ready" nodes.
-  "${KUBE_ROOT}/cluster/kubectl.sh" get nodes > "${MINIONS_FILE}" || true
-  found=$(cat "${MINIONS_FILE}" | sed '1d' | grep -c .) || true
-  ready=$(cat "${MINIONS_FILE}" | sed '1d' | awk '{print $NF}' | grep -c '^Ready') || true
+  "${KUBE_ROOT}/cluster/kubectl.sh" get nodes --api-version="${API_VERSION}" -o yaml> "${MINIONS_FILE}" || true
+  
+  found=0;ready=0
+  if [[ $API_VERSION == v1 ]]; then
```
Comment (roberthbailey) [inline cluster/validate-cluster.sh]: What do we do if the api version isn't v1? We just skip these extra checks?

File: cluster/validate-cluster.sh [hunk 6]
```diff
@@ -57,35 +61,35 @@ while true; do
...
+  componentstatuses=1;healthy=0
+  if [[ $API_VERSION == v1 ]]; then
+    # when there is a 'kind: ComponentStatus' line, there is a ComponentStatus found.
+    componentstatuses=$(echo "${kubectl_output}" | grep -c 'kind: ComponentStatus') || true
+    # only when status: "True" and type: Healthy both exist in conditions section, the compoment is "healthy".
+    healthy=$(echo "${kubectl_output}" | sed -n 'N;/status: "True".*type: Healthy/p; D;' | grep -c 'True') || true
+  fi
 
-  if ((non_success_count > 0)); then
+  if ((componentstatuses > healthy)); then
```
Comment (roberthbailey) [inline cluster/validate-cluster.sh]: I think that this will never work if api version != v1.

**Comment by roberthbailey** [comment]
@fabioy - can you take a look at the changes to validate-cluster.sh as well?

**Comment by roberthbailey** [comment]
Also needs a rebase.

File: cluster/validate-cluster.sh (hunk 3 above)
Comment (hurf) [inline cluster/validate-cluster.sh]: Oh I didn't know about the transition, can you give me some hits?

File: cluster/validate-cluster.sh (hunk 4 above)
Comment (hurf) [inline cluster/validate-cluster.sh]: Unfortunately I don't have a mac environment, so I just tested it under linux. If you or anybody else can help to test the sed command under mac environment, that would be great.

File: cluster/validate-cluster.sh (hunk 5 above)
Comment (hurf) [inline cluster/validate-cluster.sh]: I think if api version isn't v1, we need to define other checks. Currently it's set to v1.

File: cluster/validate-cluster.sh (hunk 6 above)
Comment (hurf) [inline cluster/validate-cluster.sh]: There's a value initial of `componentstatuses` and `healthy`. It will not pass the check unless these two values are handled properly for each version. Currently for v1, it works. For the coming v2 or else, we need to define a different handle process for them other than for v1.

**Comment by hurf** [comment]
@roberthbailey - Thanks for your review. And I'll try to fix the kubectl usage problem if you can give me more clues.

**Comment by roberthbailey** [comment]
@hurf This has been marked as 1.0-post so I'll get back to you with comments after the 1.0 release.

**Comment by hurf** [comment]
I thought @bgrant0607 once said want it be in 1.0.

**Comment by hurf** [comment]
Ah, just saw it, it's modified to 1.0-post yesterday.

**Comment by bgrant0607** [comment]
@hurf I did want it, but the window has passed.

**Comment by roberthbailey** [comment]
@hurf can you rebase and then I'll take another look. Thanks.

File: cluster/validate-cluster.sh (hunk 4 above)
Comment (bgrant0607) [inline cluster/validate-cluster.sh]: Does this change need to be in this PR? If not, I recommend reverting it.

File: cluster/validate-cluster.sh (hunk 6 above)
Comment (bgrant0607) [inline cluster/validate-cluster.sh]: Again, I don't see why this change needs to be in this PR.

**Comment by bgrant0607** [comment]
It was good to meet you at the launch, @hurf. Please rebase. We'd like to merge PRs that accumulated during the freeze.

File: cluster/validate-cluster.sh (hunk 4 above)
Comment (hurf) [inline cluster/validate-cluster.sh]: I've changed the output column, so if this change is not made, it never passes the test.

File: cluster/validate-cluster.sh (hunk 6 above)
Comment (hurf) [inline cluster/validate-cluster.sh]: This change is not necessary in this PR. Just a 'by the way' work.

**Comment by hurf** [comment]
@roberthbailey - rebased, can you take another look of the script to help improve it?

**Comment by hurf** [comment]
@roberthbailey - can you give some advices on the script?

**Comment by hurf** [comment]
@andronat  @jlowdermilk @thockin @erictune Can you look at this patch? It has been there for too long.

File: cluster/validate-cluster.sh (hunk 3 above)
Comment (jlowdermilk) [inline cluster/validate-cluster.sh]: I think it's fine to keep using `kubectl.sh`, as it's still used by other scripts under `cluster/`. We will transition to using kubectl directly, but find/replace isn't hard.

File: cluster/validate-cluster.sh (hunk 5 above)
Comment (jlowdermilk) [inline cluster/validate-cluster.sh]: I would say assume the api version is v1. We'll be supporting v1 for the next year or so, which may end up being longer than the bash deployment scripts last.

File: cluster/validate-cluster.sh (hunk 4 above)
Comment (jlowdermilk) [inline cluster/validate-cluster.sh]: Why not use `kubectl get -o template` here? That's pretty much what it's designed for, extracting specific values from the api object. The following command will extract `type` and `status` from each node's status.conditions[0] and output them as a comma-separated string:
```
$ kubectl get nodes -o template --template='{{range .items}}{{with index .status.conditions 0}}{{.type}}:{{.status}},{{end}}{{end}}'
Ready:True,Ready:True,%
```

File: cluster/validate-cluster.sh (hunk 6 above)
Comment (jlowdermilk) [inline cluster/validate-cluster.sh]: If you are going to add this change, I again would recommend using `-o template` to extract the values:
```
$ kubectl get componentstatuses -o template --template='{{range .items}}{{with index .conditions 0}}{{.type}}:{{.status}},{{end}}{{end}}' --api-version=v1
Healthy:True,Healthy:True,Healthy:True,
```

**Comment by jlowdermilk** [comment]
@hurf, I left you some comments. Couple things to clean up, but this should be ready to merge shortly.

File: pkg/kubectl/cmd/cmd_test.go [hunk 7]
```diff
@@ -24,6 +24,7 @@ import (
 	"io/ioutil"
 	"os"
 	"reflect"
+	"strings"
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/cmd_test.go]: Unused import. This is what's causing current shippable failure.

File: cluster/validate-cluster.sh (hunk 4 above)
Comment (hurf) [inline cluster/validate-cluster.sh]: The reason I didn't use `-o template` is when template format changes, it will not working. But as you said, the v1 version may last longer than the script, it should not be a concern. I'll change it to `-o template`

Comment (bgrant0607) [inline cluster/validate-cluster.sh]: I recommend using `-o template`, but also specify the api version with --api-version=v1.

**Comment by jlowdermilk** [comment]
lgtm, thanks for your patience, @hurf
